storm-client/src/jvm/org/apache/storm/messaging/netty/KerberosSaslClientHandler.java

```diff
@@ -96,53 +96,69 @@ private void handleControlMessage(ChannelHandlerContext ctx, ControlMessage cont
         Channel channel = ctx.channel();
         KerberosSaslNettyClient saslNettyClient = getChannelSaslClient(channel);
-            LOG.debug("Responding to server's token of length: {}",
-            saslMessageToken.getSaslToken().length);
-
-            // Generate SASL response (but we only actually send the response if
-            // it's non-null.
-            byte[] responseToServer = saslNettyClient
-            .saslResponse(saslMessageToken);
-            if (responseToServer == null) {
-                // If we generate a null response, then authentication has completed
-                // (if not, warn), and return without sending a response back to the
-                // server.
-                LOG.debug("Response to server is null: authentication should now be complete.");
-                if (!saslNettyClient.isComplete()) {
-                    LOG.warn("Generated a null response, but authentication is not complete.");
-                    throw new Exception("Our reponse to the server is null, but as far as we can tell, we are not authenticated yet.");
-                }
-            this.client.channelReady(channel);
-            } else {
-                LOG.debug("Response to server token has length: {}",
-                          responseToServer.length);
-            // Construct a message containing the SASL response and send it to the
+        LOG.debug("Responding to server's token of length: {}", saslMessageToken.getSaslToken().length);
+
+        // Generate SASL response (but we only actually send the response if
+        // it's non-null.
+        byte[] responseToServer = saslNettyClient.saslResponse(saslMessageToken);
+        if (responseToServer == null) {
+            // If we generate a null response, then authentication has completed
+            // (if not, warn), and return without sending a response back to the
             // server.
+            LOG.debug("Response to server is null: authentication should now be complete.");
+            if (!saslNettyClient.isComplete()) {
+                LOG.warn("Generated a null response, but authentication is not complete.");
+                throw new Exception("Our reponse to the server is null, but as far as we can tell, we are not authenticated yet.");
+            }
+            this.client.channelReady(channel);
+        } else {
+            LOG.debug("Response to server token has length: {}",
+                      responseToServer.length);
+            // Construct a message containing the SASL response and send it to the server.
             SaslMessageToken saslResponse = new SaslMessageToken(responseToServer);
             channel.writeAndFlush(saslResponse, channel.voidPromise());
         }
     }
+
+    @Override
+    public void channelRegistered(ChannelHandlerContext ctx) throws Exception {
+        super.channelRegistered(ctx);
+        LOG.debug("channelRegistered {}", ctx);
```

Comment:
Sorry, didn't get to this in time. It looks good overall, but I'm wondering what these methods are for? If you want to monitor the reconnect, I think channelActive/channelInactive would be a better fit (see https://netty.io/wiki/new-and-noteworthy-in-4.0.html#simplified-channel-state-model)

Reply:
In my case, the channel never went active.  I did not understand what was happening at the time.  By overriding the registered and unregistered, my first clue was that unregister was called immediately after register.  

Since it took me a day and a half to debug this issue and this was my first real clue, I thought it was important to log to save time for debugging future issues.

I will say I do not know anything about Netty though, I am sure you could have solve this much quicker.

Reply:
Thanks for explaining. Happy you solved it.